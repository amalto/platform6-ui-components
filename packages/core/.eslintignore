dist/
loader/

src/components.d.ts