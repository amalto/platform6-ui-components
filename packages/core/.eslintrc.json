{
  "extends": ["../../.eslintrc.json", "plugin:@stencil/recommended", "plugin:jest/recommended", "plugin:jsx-a11y/recommended"],
  "parserOptions": {
    "project": ["./tsconfig.json", "./tsconfig.lint.json"],
    "ecmaFeatures": {
      "jsx": true // Allows for the parsing of JSX
    }
  },
  "plugins": ["jest"],

  "rules": {
    "jsx-a11y/label-has-associated-control": [
      2,
      {
        "assert": "either"
      }
    ],
    "react/jsx-props-no-spreading": "off",
    "react/no-unknown-property": [2, { "ignore": ["class"] }]
  },
  "settings": {
    "react": {
      "pragma": "h"
    }
  }
}
