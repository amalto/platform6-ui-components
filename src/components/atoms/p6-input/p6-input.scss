@import 'src/styles/atoms/control';

:host {
  box-sizing: border-box;
  display: block;
  position: relative;

  .control {
    @include control-container;

    display: flex;
  }

  .input {
    @include input-textarea;
    &.is-static {
      background-color: transparent;
      border-color: transparent;
      box-shadow: none;
      padding-left: 0;
      padding-right: 0;
    }
  }
}
