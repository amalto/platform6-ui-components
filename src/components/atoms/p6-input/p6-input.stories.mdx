import {
  Story,
  Preview,
  Meta,
  Markdown,
  Description,
} from "@storybook/addon-docs/blocks";

import Readme from "./readme.md";

import {
  DefaultStory,
  TextStory,
  SizeStory,
  ReadonlyStory,
  DisabledStory,
  WaitingStory,
  OnErrorStory,
} from "./p6-input.stories.ts";

<Meta
  title="Atoms/Input"
  component="p6-input"
  parameters={{
    actions: {
      handles: ["p6Submit p6-form"],
    },
  }}
/>

# Input

<Preview>
  <Story
    name="Default"
    height="84px"
    parameters={DefaultStory.parameters}
    args={DefaultStory.args}
    argTypes={DefaultStory.argTypes}
  >
    {DefaultStory}
  </Story>
</Preview>

<Preview>
  <Story
    name="Text"
    height="84px"
    parameters={TextStory.parameters}
    args={TextStory.args}
    argTypes={TextStory.argTypes}
  >
    {TextStory}
  </Story>
</Preview>

<Preview>
  <Story
    name="Read only"
    height="84px"
    parameters={ReadonlyStory.parameters}
    args={ReadonlyStory.args}
    argTypes={ReadonlyStory.argTypes}
  >
    {ReadonlyStory}
  </Story>
</Preview>

<Preview>
  <Story
    name="Disabled"
    height="84px"
    parameters={DisabledStory.parameters}
    args={DisabledStory.args}
    argTypes={DisabledStory.argTypes}
  >
    {DisabledStory}
  </Story>
</Preview>

<Preview>
  <Story
    name="Waiting"
    height="84px"
    parameters={WaitingStory.parameters}
    args={WaitingStory.args}
    argTypes={WaitingStory.argTypes}
  >
    {WaitingStory}
  </Story>
</Preview>

<Preview>
  <Story
    name="On error"
    height="84px"
    parameters={OnErrorStory.parameters}
    args={OnErrorStory.args}
    argTypes={OnErrorStory.argTypes}
  >
    {OnErrorStory}
  </Story>
</Preview>

<Preview>
  <Story
    name="Size"
    height="84px"
    parameters={SizeStory.parameters}
    args={SizeStory.args}
    argTypes={SizeStory.argTypes}
  >
    {SizeStory}
  </Story>
</Preview>

<Description markdown={Readme} />
