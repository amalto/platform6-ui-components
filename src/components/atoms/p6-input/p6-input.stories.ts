import { Components } from '../../../components';
import { ComponentProps, getElement, getForm, getSelectArgType, makeStory, Props } from '../../../shared/storybook';
import { enumToArray } from '../../../utils/enum';
import { P6InputType } from './types';

const component = 'p6-input';

export default {
  title: 'Atoms/Input',
  component,
};

const componentProps: ComponentProps = ['disabled', 'max', 'min', 'name', 'pattern', 'placeholder', 'readOnly', 'required', 'type', 'value', 'waiting'];

const getStoryField = (props?: Props<Components.P6Input>): HTMLElement => getElement(component, [], props);

export const Default = makeStory<{
  disabled: boolean;
  readOnly: boolean;
  required: boolean;
  waiting: boolean;
  min: number;
  max: number;
  pattern: string;
  placeholder: string;
  type: P6InputType;
  value: string;
}>({
  componentProps,
  args: {
    disabled: false,
    readOnly: false,
    required: false,
    waiting: false,
    min: 0,
    max: 42,
    pattern: '',
    placeholder: 'Placeholder',
    type: P6InputType.text,
    value: '',
  },
  argTypes: {
    ...getSelectArgType('type', enumToArray(P6InputType)),
  },
  builder: (args): HTMLElement => getStoryField(args),
});

export const Text = makeStory({
  componentProps,
  builder: (): HTMLElement => getStoryField({ type: P6InputType.text }),
});

export const Readonly = makeStory<{ readOnly: boolean }>({
  componentProps,
  args: { readOnly: true },
  builder: (props): HTMLElement => getStoryField({ placeholder: 'Read only', ...props }),
});

export const Disabled = makeStory<{ disabled: boolean }>({
  componentProps,
  args: { disabled: true },
  builder: (props): HTMLElement => getStoryField({ placeholder: 'Disabled', ...props }),
});

export const Waiting = makeStory<{ waiting: boolean }>({
  componentProps,
  args: { waiting: true },
  builder: (props): HTMLElement => getStoryField({ placeholder: 'Waiting', ...props }),
});

export const OnError = makeStory({
  componentProps,
  builder: (): HTMLElement => getStoryField({ pattern: '42', value: '84' }),
});

export const Form = makeStory<{
  disabled: boolean;
  readOnly: boolean;
  required: boolean;
  type: P6InputType;
  value: string;
}>({
  componentProps,
  args: {
    disabled: false,
    readOnly: false,
    required: false,
    type: P6InputType.text,
    value: 'default value',
  },
  argTypes: {
    ...getSelectArgType('type', enumToArray(P6InputType)),
  },
  builder: (props): HTMLElement => getForm(getStoryField({ name: 'field', ...props })),
});
