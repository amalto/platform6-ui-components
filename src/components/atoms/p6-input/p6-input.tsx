import {
  Component,
  Element,
  h,
  Host,
  Method,
  Prop,
  State,
} from "@stencil/core";
import { cleanupValue, isEmpty } from "~utils/attribute";

export type P6InputType =
  | "email"
  | "number"
  | "password"
  | "search"
  | "tel"
  | "text"
  | "url";

@Component({
  tag: "p6-input",
  styleUrl: "p6-input.scss",
  scoped: true,
})
export class P6Input {
  @Element() host!: HTMLP6InputElement;

  /**
   * the input is not available for interaction. The value will not be submitted with the form
   */
  @Prop() disabled = false;

  /**
   * The maximum length or value
   */
  @Prop() max: number | undefined;

  /**
   * The minimum length or value
   */
  @Prop() min: number | undefined;

  /**
   * The name of the input.
   */
  @Prop() name!: string;

  /**
   * Pattern the value must match to be valid.
   */
  @Prop() pattern: string | undefined;

  /**
   * content to be appear in the form control when the form control is empty
   */
  @Prop() placeholder = "";

  /**
   * marks an element that can't be edited.
   */
  @Prop({ attribute: "readOnly" }) readOnly = false;

  /**
   * marks an element that can't be submitted without a value.
   */
  @Prop() required = false;

  /**
   * the content type of the input.
   */
  @Prop() type: P6InputType = "text";

  /**
   * the value of the input.
   */
  @Prop() value: string | undefined;

  /**
   * shows a waiting indicator
   */
  @Prop() waiting = false;

  @State() hasError = false;

  private nativeInput: HTMLInputElement | undefined;

  componentWillLoad(): void {
    this.host.addEventListener(
      "focusout",
      this.internalCheckValidity.bind(this)
    );
  }

  componentDidLoad(): void {
    if (!isEmpty(this.value)) {
      this.internalCheckValidity();
    }
  }

  render(): JSX.Element {
    const containerClass = {
      "is-loading": !!this.waiting,
    };

    const classes = {
      input: true,
      "is-danger": this.hasError,
      "is-static": !!this.readOnly,
    };

    return (
      <Host class={containerClass}>
        <input
          class={classes}
          ref={(input): void => {
            this.nativeInput = input;
          }}
          disabled={this.disabled}
          name={this.name}
          pattern={cleanupValue(this.pattern)}
          placeholder={cleanupValue(this.placeholder)}
          readOnly={this.readOnly}
          required={this.required}
          type={this.type}
          value={this.value}
          {...this.minMaxAttrs}
        />
      </Host>
    );
  }

  /**
   * Returns the error message that would be displayed if the user submits the form, or an empty string if no error message.
   * It also triggers the standard error message, such as "this is a required field".
   */
  @Method()
  async validationMessage(): Promise<string> {
    return Promise.resolve(this.nativeInput?.validationMessage || "");
  }

  /**
   * Returns whether a form will validate when it is submitted, without having to submit it.
   */
  @Method()
  async checkValidity(): Promise<boolean> {
    return Promise.resolve(this.nativeInput?.checkValidity() || true);
  }

  private get minMaxAttrs():
    | { min?: number; max?: number }
    | { minLength?: number; maxLength?: number } {
    return this.type === "number"
      ? {
          min: this.min,
          max: this.max,
        }
      : {
          minLength: this.min,
          maxLength: this.max,
        };
  }

  private async internalCheckValidity(): Promise<void> {
    await this.checkValidity();
    const msg = await this.validationMessage();
    this.hasError = msg !== "";
  }
}
