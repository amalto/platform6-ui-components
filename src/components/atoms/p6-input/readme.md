# p6-input

<!-- Auto Generated Below -->


## Properties

| Property            | Attribute     | Description                                                                               | Type                                                                        | Default     |
| ------------------- | ------------- | ----------------------------------------------------------------------------------------- | --------------------------------------------------------------------------- | ----------- |
| `disabled`          | `disabled`    | the input is not available for interaction. The value will not be submitted with the form | `boolean`                                                                   | `false`     |
| `max`               | `max`         | The maximum length or value                                                               | `string \| undefined`                                                       | `undefined` |
| `min`               | `min`         | The minimum length or value                                                               | `string \| undefined`                                                       | `undefined` |
| `multiline`         | `multiline`   | Enables multiline support (with a textarea instead of an input)                           | `boolean`                                                                   | `false`     |
| `name` _(required)_ | `name`        | The name of the input.                                                                    | `string`                                                                    | `undefined` |
| `pattern`           | `pattern`     | Pattern the value must match to be valid.                                                 | `string \| undefined`                                                       | `undefined` |
| `placeholder`       | `placeholder` | content to be appear in the form control when the form control is empty                   | `string \| undefined`                                                       | `undefined` |
| `readonly`          | `readonly`    | marks an element that can't be edited.                                                    | `boolean`                                                                   | `false`     |
| `required`          | `required`    | marks an element that can't be submitted without a value.                                 | `boolean`                                                                   | `false`     |
| `type`              | `type`        | the content type of the input.                                                            | `"email" \| "number" \| "password" \| "search" \| "tel" \| "text" \| "url"` | `"text"`    |
| `value`             | `value`       | the value of the input.                                                                   | `string \| undefined`                                                       | `undefined` |
| `waiting`           | `waiting`     | shows a waiting indicator                                                                 | `boolean`                                                                   | `false`     |


## Methods

### `checkValidity() => Promise<boolean>`

Returns whether a form will validate when it is submitted, without having to submit it.

#### Returns

Type: `Promise<boolean>`




----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
