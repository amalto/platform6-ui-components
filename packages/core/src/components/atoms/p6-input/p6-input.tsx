import { Component, ComponentInterface, Element, Event, EventEmitter, h, Host, JSX, Method, Prop, State } from '@stencil/core';
import { P6Control } from '../../../shared/form/control';
import { InvalidEventDetail, ValidEventDetail } from '../../../shared/form/event';
import { defaultCheckValidity, defaultValidationMessage } from '../../../shared/form/validation';
import { cleanupValue, isEmpty } from '../../../utils/attribute';
import { P6InputType } from './types';

export type P6InputValue = string | number | undefined;

@Component({
  tag: 'p6-input',
  styleUrl: 'p6-input.scss',
  shadow: true,
})
export class P6Input implements ComponentInterface, P6Control<P6InputValue> {
  @Element() host!: HTMLP6InputElement;

  /**
   * the input is not available for interaction. The value will not be submitted with the form
   */
  @Prop() disabled = false;

  /**
   * The maximum length or value
   */
  @Prop() max: number | undefined;

  /**
   * The minimum length or value
   */
  @Prop() min: number | undefined;

  /**
   * The name of the input.
   */
  @Prop() name!: string;

  /**
   * Pattern the value must match to be valid.
   */
  @Prop() pattern: string | undefined;

  /**
   * content to be appear in the form control when the form control is empty
   */
  @Prop() placeholder = '';

  /**
   * marks an element that can't be edited.
   */
  @Prop({ attribute: 'readOnly' }) readOnly = false;

  /**
   * marks an element that can't be submitted without a value.
   */
  @Prop() required = false;

  /**
   * the content type of the input.
   */
  @Prop() type: P6InputType = P6InputType.text;

  /**
   * the value of the input.
   */
  @Prop() value: string | number | undefined;

  /**
   * shows a waiting indicator
   */
  @Prop() waiting = false;

  @State() hasError = false;

  /**
   * Registering the field in a p6-form
   */
  @Event() p6FormRegister!: EventEmitter<P6Control<P6InputValue>>;

  /**
   * Unregistering the field in a p6-form
   */
  @Event() p6FormUnregister!: EventEmitter<P6Control<P6InputValue>>;

  /**
   * Fires when the field has been checked for validity and satisfy its constraints
   */
  @Event() p6Valid!: EventEmitter<ValidEventDetail<P6InputValue>>;

  /**
   * Fires when the field has been checked for validity and doesn't satisfy its constraints
   */
  @Event() p6Invalid!: EventEmitter<InvalidEventDetail>;

  private nativeInput: HTMLInputElement | undefined;

  private defaultValue: string | number | undefined;

  componentWillLoad(): void {
    this.host.addEventListener('focusout', this.checkValidity.bind(this));

    if (this.defaultValue === undefined) {
      this.defaultValue = this.value;
    }
  }

  render(): JSX.Element {
    const containerClass = {
      'control': true,
      'is-loading': !!this.waiting,
    };

    const classes = {
      'input': true,
      'is-danger': this.hasError,
      'is-static': !!this.readOnly,
    };

    return (
      <Host>
        <form noValidate class={containerClass}>
          <input
            class={classes}
            ref={(input): void => {
              this.nativeInput = input;
            }}
            disabled={this.disabled}
            name={this.name}
            pattern={cleanupValue(this.pattern)}
            placeholder={cleanupValue(this.placeholder)}
            readOnly={this.readOnly}
            required={this.required}
            type={P6InputType[this.type]}
            value={this.value}
            {...this.minMaxAttrs}
          />
        </form>
      </Host>
    );
  }

  async componentDidLoad(): Promise<void> {
    if (!isEmpty(this.value)) {
      await this.checkValidity();
    }

    this.p6FormRegister.emit(this);
  }

  disconnectedCallback?(): void {
    this.p6FormUnregister.emit(this);
  }

  /**
   * Returns the error message that would be displayed if the user submits the form, or an empty string if no error message.
   * It also triggers the standard error message, such as "this is a required field".
   */
  @Method()
  async validationMessage(): Promise<string> {
    return defaultValidationMessage(this.nativeInput);
  }

  /**
   * Returns whether a form will validate when it is submitted, without having to submit it.
   */
  @Method()
  async checkValidity(): Promise<boolean> {
    return defaultCheckValidity<P6InputValue>({
      name: this.name,
      disabled: this.disabled,
      nativeInput: this.nativeInput,
      p6Valid: this.p6Valid,
      p6Invalid: this.p6Invalid,
      validationMessage: this.validationMessage.bind(this),
      errorHandler: hasError => {
        this.hasError = hasError;
      },
      getValue: () => this.inputValue,
    });
  }

  /**
   * Restores the input's default value
   */
  @Method()
  async reset(): Promise<boolean> {
    let result = false;

    if (this.nativeInput) {
      this.nativeInput.value = this.defaultValue !== undefined ? this.defaultValue.toString() : '';
      result = true;
    }

    return Promise.resolve(result);
  }

  private get minMaxAttrs(): { min?: number; max?: number } | { minLength?: number; maxLength?: number } {
    return this.type === P6InputType.number
      ? {
          min: this.min,
          max: this.max,
        }
      : {
          minLength: this.min,
          maxLength: this.max,
        };
  }

  private get inputValue(): string | number | undefined {
    if (this.nativeInput === undefined) {
      return undefined;
    }

    if (this.nativeInput.type === P6InputType[P6InputType.number]) {
      return this.nativeInput.valueAsNumber;
    }

    return this.nativeInput.value;
  }
}
