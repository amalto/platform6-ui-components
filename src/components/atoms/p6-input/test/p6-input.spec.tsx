import { newSpecPage } from "@stencil/core/testing";
import { P6Input } from "../p6-input";

describe("p6-input", () => {
  Object.defineProperty(HTMLInputElement.prototype, "checkValidity", {
    writable: true,
    value: jest.fn().mockImplementation(() => true)
  });

  it("renders", async () => {
    const page = await newSpecPage({
      components: [P6Input],
      html: `<p6-input>label</p6-input>`
    });
    expect(page.root).toEqualHtml(`
      <p6-input>
        <label class="label" htmlfor="undefined-input">label</label>
        <div class="control">
          <input class="input" id="undefined-input" type="text" />
        </div>
      </p6-input>
    `);
  });

  it("add the waiting class on the input container, not on the input", async () => {
    const page = await newSpecPage({
      components: [P6Input],
      html: `<p6-input waiting>label</p6-input>`
    });
    expect(page.root).toEqualHtml(`
      <p6-input waiting>
        <label class="label" htmlfor="undefined-input">label</label>
        <div class="control is-loading">
          <input class="input" id="undefined-input" type="text" />
        </div>
      </p6-input>
    `);
  });

  it('add the class "is-static" when the readonly props is true', async () => {
    const page = await newSpecPage({
      components: [P6Input],
      html: `<p6-input readonly>label</p6-input>`
    });

    expect(page.root).toEqualHtml(`
      <p6-input readonly>
        <label class="label" htmlfor="undefined-input">label</label>
        <div class="control">
          <input class="input is-static" id="undefined-input" type="text" readonly />
        </div>
      </p6-input>
    `);
  });
});
