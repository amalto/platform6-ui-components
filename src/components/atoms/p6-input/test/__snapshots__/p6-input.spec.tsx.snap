// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`p6-input add the class "is-static" when the readonly props is true 1`] = `
<p6-input readonly="">
  <mock:shadow-root>
    <div class="control">
      <input class="input is-static" readonly="" type="text">
    </div>
  </mock:shadow-root>
</p6-input>
`;

exports[`p6-input add the waiting class on the input container, not on the input 1`] = `
<p6-input waiting="">
  <mock:shadow-root>
    <div class="control is-loading">
      <input class="input" type="text">
    </div>
  </mock:shadow-root>
</p6-input>
`;

exports[`p6-input renders 1`] = `
<p6-input>
  <mock:shadow-root>
    <div class="control">
      <input class="input" type="text">
    </div>
  </mock:shadow-root>
</p6-input>
`;
